src/k2/tso/client/tso_clientlib.cpp

```diff
@@ -87,42 +87,37 @@ seastar::future<> TSO_ClientLib::DiscoverServiceNodes(const k2::String& serverUR
+        GetTSOServiceNodeURLsRequest request;  // empty request param
+
+        return k2::RPC().callRPC<dto::GetTSOServiceNodeURLsRequest, dto::GetTSOServiceNodeURLsResponse>(dto::Verbs::GET_TSO_SERVICE_NODE_URLS, request, *myRemote, timeout)
+        .then([this](auto&& response) {
             if (_stopped) return seastar::make_ready_future<>();
 
-            if (!payload || payload->getSize() == 0)
+            auto& [status, r] = response;
+            if (!status.is2xxOK())
             {
-                K2LOG_E(log::tsoclient, "Remote end did not provide a data endpoint. Giving up");
-                return seastar::make_exception_future<>(std::runtime_error("no remote endpoint"));
+                K2LOG_E(log::tsoclient, "Error during get TSO node URLs, code:{}, error message:{}", status.code, status.message);
+                // currently, it is not expected 
+                return seastar::make_exception_future<>(std::runtime_error(status.message));
             }
 
-            std::vector<std::vector<k2::String>> nodeURLs;
-            payload->read(nodeURLs);
-            K2ASSERT(log::tsoclient, !nodeURLs.empty(), "TSO server should have workers");
+            auto& nodeURLs = r.serviceNodeURLs;
+            if (nodeURLs.empty())
+            {
+                K2LOG_E(log::tsoclient, "Remote end did not provide node URLs. Giving up");
+                return seastar::make_exception_future<>(std::runtime_error("no remote endpoint"));
+            }
+            else
+            {
+                K2LOG_I(log::tsoclient, "received node URLs:{}", nodeURLs);
+            }
 
             _curTSOServiceNodes.clear();
             // each node may have mulitple endPoints URLs, we only pick the fastest supported one, currently RDMA, if no RDMA, pick TCPIP
             for (auto& singleNodeURLs : nodeURLs)
             {
-                k2::TXEndpoint endPointToAdd;
-                for (auto& url : singleNodeURLs)
-                {
-                    auto tempEndPoint = *(k2::RPC().getTXEndpoint(url));
-                    K2LOG_I(log::tsoclient, "Found remote data endpoint: {}", url);
-                    if (tempEndPoint.protocol == RRDMARPCProtocol::proto)
-                    {
-                        // if found RDMA, use it and break out
-                        endPointToAdd = tempEndPoint;
-                        break;
-                    }
-                    else if (tempEndPoint.protocol == TCPRPCProtocol::proto)
-                    {
-                        // keep it to enPointToAdd, maybe replaced by RDMA endpoint later
-                        endPointToAdd = tempEndPoint;
-                    }
-                }
-                _curTSOServiceNodes.emplace_back(endPointToAdd);
+                _curTSOServiceNodes.push_back( Discovery::selectBestEndpoint(singleNodeURLs));
+                K2LOG_I(log::tsoclient, "Selected node endpoint:{}", _curTSOServiceNodes.back()->url);

```

Comment:
probably just debug here too

Reply:
This is during initial setup, normally once per lift time. So far for such initialization, we have been using INFO level tracing, so I will keep it as it is now.